perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/AppAutoRejectionScheduler.java

```diff
@@ -0,0 +1,252 @@
+			}
+		}
+	}
+
+	/**
+	 * Compares date of last modification of application to values in expiration attribute and if finds expired application, then
+	 * rejects it.
+	 *
+	 * @param applications applications
+	 * @param expiration attribute with number of days to application expiration
+	 */
+	private void rejectExpiredApplications (List<Application> applications, Attribute expiration) {
+		Map<String, String> attrValue = expiration.valueAsMap();
+		for(Application application : applications) {
+			String date = application.getModifiedAt();
+			LocalDate modifiedAt = LocalDate.parse(date.substring(0, 10));
+			LocalDate now = getCurrentLocalDate();
+			if (application.getState() == Application.AppState.NEW) {
+				int expirationAppWaitingForEmail = Integer.parseInt(attrValue.get("emailVerification"));
+				if (now.minusDays(expirationAppWaitingForEmail).isAfter(modifiedAt)) {
+					String reasonForVo = "Your application to VO " + application.getVo().getName() + " was auto rejected, because you didn't verify your email address.";
+					String reasonForGroup = application.getGroup() == null ? "" : "Your application to group " + application.getGroup().getName() + " was auto rejected, because you didn't verify your email address.";
+					rejectWithReason(application, reasonForVo, reasonForGroup);
+					continue;
+				}
+			}
+			int expirationAppIgnoredByAdmin = Integer.parseInt(attrValue.get("ignoredByAdmin"));
+			if (now.minusDays(expirationAppIgnoredByAdmin).isAfter(modifiedAt)) {
+				String reasonForVo = "Your application to VO " + application.getVo().getName() + " was auto rejected, because admin didn't approve your application in a timely manner.";
+				String reasonForGroup = application.getGroup() == null ? "" : "Your application to group " + application.getGroup().getName() + " was auto rejected, because admin didn't approve your application in a timely manner.";
+				rejectWithReason(application, reasonForVo, reasonForGroup);
+			}
+		}
+	}
+
+	/**
+	 * Rejects given application to Vo or group due to given reason.
+	 *
+	 * @param application application to reject
+	 * @param reasonForVo reason for reject VO application
+	 * @param reasonForGroup reason for reject group application
+	 */
+	private void rejectWithReason (Application application, String reasonForVo, String reasonForGroup) {
+		try {
+			if (application.getGroup() == null) {
+				registrarManager.rejectApplication(sess, application.getId(), reasonForVo);
+			} else {
+				registrarManager.rejectApplication(sess, application.getId(), reasonForGroup);
+			}
+		} catch (PerunException e) {
+			log.error("Failed to reject expired application: {}", application, e);
+		}
+
+	}
+
+	/**
+	 * Selects all vos from database, in which could be some expired applications acceptable for auto rejection.
+	 *
+	 * @return list of vos
+	 * @throws VoNotExistsException if vo not exist (it shouldn't happen)
+	 */
+	private List<Vo> getAllEligibleVos() throws VoNotExistsException {
```

Comment:
I would suggest changing the return type of these methods to `List<Integer>` and moving them to the SearcherImpl component. Then you can call here the SearcherBl and map the ids to Vos.

Reply:
Query moved to the SearcherImpl.